### DISCUSSION TITLE
[RFC] Environment Variable Loading

### DISCUSSION THREAD

**Discussion by flybayer** [blitz-js/blitz#2444] [RFCs]
## Problem

Blitz/Next use basically [dotenv-flow](https://github.com/kerimdzhanov/dotenv-flow) to load env var files like `.env`, `.env.test`, `.env.development.local`, etc. 

That behavior relies on setting NODE_ENV to the appropriate value before invoking a CLI command.

Problems:

1. Nextjs internals don't allow changing NODE_ENV. The `build` command must use `NODE_ENV=production` and `dev` must use `NODE_ENV=development`.
2. There's a [new Nextjs bug](https://github.com/vercel/next.js/issues/25491) that breaks previous usage of `NODE_ENV=test blitz build`
3. Now it's impossible to use `NODE_ENV` inside your app code to conditionally do things based on different environments.

## Proposed Solution

Use `APP_ENV` to determine app environment instead of `NODE_ENV`. `NODE_ENV` will only be used for `blitz dev` vs `blitz build`.

The value of `APP_ENV` will determine which `.env.[ENV].*` files will be loaded.

1. Defaults:
   1. `APP_ENV=development` for `blitz dev` 
   2. `APP_ENV=production` for `blitz build` and `blitz start`
   3. `APP_ENV=test` for Jest tests
2. Ways to manually specify APP_ENV:
   1. Set it before invoking the CLI like `APP_ENV=test blitz start`
   2. Set it with a new `-e` flag like `blitz start -e test`

With this solution, you can set `APP_ENV` to any string. Example: `APP_ENV=staging` which would load `.env.staging`

---

Any and all feedback is welcome!

**Comment by lksnmnn** [comment]
Would that solution help us to make the ENV loading work better together with prisma? In our deployment, we currently overwrite `.env` to make it work.

**Reply by flybayer** [reply]
The existing solution for Prisma is to always use `blitz prisma COMMAND` instead of `prisma COMMAND`. `blitz prisma` will load all the appropriate `.env` files

**Reply by lksnmnn** [reply]
Oh, I wasn't reading up on that. Thanks! I'll give it a try.

**Comment by flybayer** [comment]
I've also posted this upstream in Next.js repo: https://github.com/vercel/next.js/discussions/25764
